manta-trusted-setup/src/ceremony/coordinator.rs

```diff
@@ -58,84 +63,79 @@ where
-
-        // make sure it is participant's turn
+            .ok_or(CeremonyError::NotRegistered)
+            .expect("Get the participant from registry should succeed.");
         if !self.queue.is_at_front(participant) {
             return Err(CeremonyError::NotYourTurn);
         };
-
-        // verify and update the state and challenge
-        let (_, transformed_state) = V::verify_transform(
-            &self.challenge,
-            core::mem::take(&mut self.state),
-            transformed_state,
-            proof,
-        )
-        .expect("Verify transform on received contribution should succeed.");
-        self.state = transformed_state;
-
-        // remove the participant from the queue but the participant
-        // will noe be removed from the registry, so the participant will not
-        // be able to participate in this ceremony again.
+        (_, self.state) = V::verify_transform(&self.challenge, take(&mut self.state), state, proof)
+            .expect("Verify transform on received contribution should succeed.");
         self.queue.pop();
-        // println!("MPC state updated");
         Ok(())
     }
 
-    /// Register a participant and put them into the waiting queue.
+    /// Registers a participant and puts into the waiting queue.
+    #[inline]
     pub fn register(&mut self, participant: P) -> Result<(), CeremonyError> {
         let participant = self
             .registry
-            .register(participant.identifier(), participant)?;
-        self.queue.push(participant);
+            .register(participant.identifier(), participant)
+            .expect("Register a participant should succeed.");
+        self.queue
+            .push(participant.priority(), participant.identifier());
         Ok(())
     }
 
-    /// Get the participant with the given identifier. Returns `None` if not found.
+    /// Gets the participant with the given identifier and returns `None` if not found.
+    #[inline]
     pub fn get_participant(&self, identifier: &P::Identifier) -> Option<&P> {
         self.registry.get(identifier)
     }
 
-    /// Put the current contributor and move to next one. Return the participant that is skipped.
-    /// The skipped participant needs to be registered again.
-    pub fn skip_current_contributor(&mut self) -> Result<P, CeremonyError> {
-        let participant = self.queue.pop().ok_or(CeremonyError::WaitingQueueEmpty)?;
-        Ok(self
-            .registry
-            .unregister(&participant)
-            .expect("participant in the queue should be registered"))
+    /// Pops the current contributor and moves to the back.
+    #[inline]
+    pub fn skip_current_contributor(&mut self) -> Result<(), CeremonyError> {
+        let (priority, identifier) = self
+            .queue
+            .pop()
+            .ok_or(CeremonyError::WaitingQueueEmpty)
+            .expect("Poping the current participant should succeed.");
+        self.queue.push(priority, identifier);
```

Comment:
What if there is only 1 person left in the queue but he never contributes? Then this is a infinite loop.